File: CodeSignal/SingleFunction2b.java
/*
 * A sawtooth sequence is a sequence of numbers that alternate between increasing and decreasing. 
 * In other words, each element is either strictly greater than its neighbouring elements or strictly less than its neighbouring elements.
 * Given an array of integers arr, your task is to count the number of contiguous subarrays that represent a sawtooth sequence of at least two elements.
 */



 /*
    Tests passed: 16/16.
    Sample tests:
    8/8
    Hidden tests:
    8/8
    Score:
    500/500
  */


  long solution(int[] arr) {
    //edge case
    if (arr.length < 2)
        return 0;
    
    //initialize variables
    long counter = 0;
    int streak = 0;
    boolean prevIncreasing = false;
    
    //loop through array
    for (int i = 0; i < arr.length-1; i++) {
        //if current element is equal to next element, reset streak and prevIncreasing
        if (arr[i] == arr[i+1]){
            prevIncreasing = false;
            streak = 0;    
        } else {
            //check if current element is increasing or decreasing
            boolean currIncreasing = arr[i] < arr[i+1];
            //if there is a flip in direction, increment streak and set prevIncreasing
            if (currIncreasing != prevIncreasing){
                streak++;
                prevIncreasing = currIncreasing;
            } else {
                //otherwise, start a new streak
                streak =1;
            }
        }
        //increment counter by the value of streak
        counter += streak;
    }   
    //return the final counter value
    return counter;
}






 
 /*
  * Tests passed: 8/16. Execution time limit exceeded on test 9: Program exceeded the execution time limit. Make sure that it completes execution in a few seconds for any possible input.
    Click the "Run Tests" button to see output, console logs, and detailed error messages for sample or custom test cases. This information is hidden when clicking the "Submit" button in order to prevent hard-coding solutions to the hidden test cases.
    Sample tests:
    8/8
    Hidden tests:
    0/8
    Score:
    167/500
  */
long solution(int[] arr)  {
    long result = 0;
    for(int start = 0; start<arr.length-1; start++) {
        int end = start+1;
        while(end<arr.length && isValid(start, end, arr)) {
            result++;
            end++;
        }
    }
    return result;
}

private boolean isValid(int start, int end, int[] arr) {
    for(int i = start+1; i<=end; i++) {
        int current = arr[i], prev = arr[i-1];
        if(i == end) {
            if(current > prev || current < prev) {
                continue;
            }  else {
                return false;
            }
        } else {
            int next = arr[i+1];
            if(current > next && current > prev || current < next && current < prev) {
                continue;
            } else {
                return false;
            }
        }
    }
    return true;
}

File: CodeSignal/SingleFunction2b.py
def solution(arr):
    if len(arr) < 2:  # Check if the array has less than 2 elements
        return 0  # Return 0 as there are no contiguous subarrays possible with less than 2 elements

    count = 0  # Initialize the count of sawtooth contiguous subarrays to 0
    streak = 0  # Initialize the streak counter to 0, which will track flips between increasing and decreasing
    prev_increasing = None  # Initialize the variable to store the direction of the previous change (increasing or decreasing)

    for i in range(1, len(arr)):  # Loop through the array starting from index 1
        if arr[i] == arr[i-1]:  # Check if the current element is equal to the previous element
            prev_increasing = None  # Reset the previous direction as the elements are the same
            streak = 0  # Reset the streak counter to 0 as there's no flip between increasing and decreasing
        else:
            curr_increasing = arr[i] > arr[i-1]  # Check if the current element is greater than the previous element
            if curr_increasing != prev_increasing:  # Check if there's a flip in direction from the previous element
                streak += 1  # Increase the streak counter as we have a flip in direction
                prev_increasing = curr_increasing  # Update the previous direction with the current direction
            else:
                streak = 1  # Reset the streak counter to 1 when we break out of a streak of flips

            count += streak  # Add the current streak count to the total count of sawtooth contiguous subarrays

    return count  # Return the final count of sawtooth contiguous subarrays


"""
Tests passed: 16/16.
Sample tests:
8/8
Hidden tests:
8/8
Score:
500/500
"""


"""
prev_increasing is used to keep track of the direction of the previous change (increase or decrease) in the array. 
It's not just a boolean; it can have three possible values: True, False, or None.

Here's how it works:

When we encounter the first element in the array, there's no previous element to compare it with, so we initialize prev_increasing to None.
As we loop through the array, we compare the current element with the previous element (arr[i] with arr[i-1]).
If the current element is greater than the previous element (arr[i] > arr[i-1]), it means we have an increasing streak, and we set curr_increasing to True.
If the current element is less than the previous element (arr[i] < arr[i-1]), it means we have a decreasing streak, and we set curr_increasing to False.
If the current element is equal to the previous element (arr[i] == arr[i-1]), there's no flip in direction, so we set curr_increasing to None.

"""